File: 022. Generate Parentheses/Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        
        self.DFS(0, 0, "", n, res)
        return res
    def DFS(self, left, right, path, n, res):
        if right > left or left > n or right > n: return;
        
        if left == n and right == n:
            
            res.append(path);
            return;   
        
        self.DFS(left + 1, right, path + '(', n, res)
        self.DFS(left, right + 1, path + ')', n, res)
    
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        dp = [[] for i in range(n + 1)]
        dp[0].append("")
        for i in range(1, n + 1):
            for k in range(i):
                dp[i] += ["(" + x + ")" + y for x in dp[k] for y in dp[i - 1 - k]]      
        
        
        return dp[n]

File: 022. Generate Parentheses/Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        
        DFS(0, 0, "", n, res);
        return res;
    }
    void DFS(int left, int right, string path, int n, vector<string>& res){
        if(right > left || left > n || right > n) return;
        
        if(left == n && right == n){
            
            res.push_back(path);
            return;   
        }
        DFS(left + 1, right, path + '(', n, res);
        DFS(left, right + 1, path + ')', n, res);
    }
    
};
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector< vector<string> > dp(n + 1, vector<string>());
        //vector<string> dp[n + 1];
        dp[0].push_back("");
        for(int i = 1; i <= n; i++){
            for(int k = 0; k < i;k++){
                for(string s2: dp[i-1-k]){
                    for(string s1: dp[k]){
                        dp[i].push_back("(" + s1 + ")" + s2);
                    }
                }
            }
        }
        return dp[n];
    }
   
    
};
